protocols/kad/src/handler.rs

```diff
@@ -950,69 +1017,160 @@ impl futures::Stream for InboundSubstreamState {
+                                    substream,
+                                    response_timeout,
+                                );
                             }
+                            // TODO: close here?
                             Err(_) => return Poll::Ready(None),
                         },
-                        Poll::Pending => {
-                            *this = InboundSubstreamState::PendingSend(id, substream, msg);
+                        (Poll::Pending, Poll::Pending) => {
+                            *this = InboundSubstreamState::PendingSend(
+                                id,
+                                substream,
+                                msg,
+                                response_timeout,
+                            );
                             return Poll::Pending;
                         }
-                        Poll::Ready(Err(_)) => return Poll::Ready(None),
+                        // TODO: close here? (x2)
+                        (Poll::Pending, Poll::Ready(())) => {
+                            tracing::debug!(
+                                connection_id = ?id,
+                                "Inbound substream timed out waiting for response send",
+                            );
+
+                            return Poll::Ready(None);
+                        }
+                        (Poll::Ready(Err(_)), _) => return Poll::Ready(None),
                     }
                 }
-                InboundSubstreamState::PendingFlush(id, mut substream) => {
-                    match substream.poll_flush_unpin(cx) {
-                        Poll::Ready(Ok(())) => {
+                InboundSubstreamState::PendingFlush(id, mut substream, mut response_timeout) => {
+                    match (
+                        substream.poll_flush_unpin(cx),
+                        response_timeout.poll_unpin(cx),
+                    ) {
+                        (Poll::Ready(Ok(())), _) => {
                             *this = InboundSubstreamState::WaitingMessage {
                                 first: false,
                                 connection_id: id,
                                 substream,
+                                request_timeout: Delay::new(SUBSTREAM_TIMEOUT),
                             };
                         }
-                        Poll::Pending => {
-                            *this = InboundSubstreamState::PendingFlush(id, substream);
+                        (Poll::Pending, Poll::Pending) => {
+                            *this = InboundSubstreamState::PendingFlush(
+                                id,
+                                substream,
+                                response_timeout,
+                            );
                             return Poll::Pending;
                         }
-                        Poll::Ready(Err(_)) => return Poll::Ready(None),
+                        // TODO: close here? (x2)
```

Comment:
make sense to close the current request here and reuse the substream for next one

Reply:
The upstream code already re-uses the substream slot here. If we are at the substream limit, the old substream ID is dropped, and a new one is used. This PR just makes more substreams available for re-use, by adding a timeout to the first request.

This TODO is about whether we need to explicitly call the `close()` method on the substream. The old code didn't, and it isn't needed to fix the bug, so I'd like to wait for a review from upstream before making this change.

It might be that `close()` does nothing, or nothing important. It's possible substreams don't have any resources, they are just an ID in a vector. And that's dropped when the limit is reached anyway.